Slang (debug) does not build with latest MSVC on Windows for versions v2026.5 or later

I'm using the latest version of Visual Studio build tools (17.14.30) on Windows. I can build releases v2026.4.2 or earlier, but staring with v2026.5 and later, debug builds from source fail. I've seen this error before and it was corrected once previously. It seems it's back. The build error is listed below:

Building Custom Rule C:/SDKBuild/build-X64-1.4.349.0/slangd/tools/CMakeLists.txt
  main.cpp
  slang-bootstrap.vcxproj -> C:\SDKBuild\build-X64-1.4.349.0\slangd\build\generators\Debug\bin\slang-bootstrap.exe
  Generating slang-core-module-generated.h, ../slang-glsl-module/slang-glsl-module-generated.h
  Compiling core module on debug build, this can take a while.
  Compiling core module took 23.00 seconds.
  Detected memory leaks!
  Dumping objects ->
  {17106561} normal block at 0x000001D9B291D7C0, 32 bytes long.
   Data: <                > 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00
  {17106560} normal block at 0x000001D9B2911110, 16 bytes long.
   Data: <  "F            > B8 ED 22 46 F8 7F 00 00 00 00 00 00 00 00 00 00
  {17106559} normal block at 0x000001D9B2911340, 16 bytes long.
   Data: <  "F            > 98 ED 22 46 F8 7F 00 00 00 00 00 00 00 00 00 00
  {650} normal block at 0x000001D9B2917800, 49 bytes long.
   Data: <   E            > 08 B1 FB 45 F8 7F 00 00 01 00 00 00 00 00 00 00
  {649} normal block at 0x000001D9B291DCA0, 32 bytes long.
   Data: <                > 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00
  {648} normal block at 0x000001D9B291A060, 16 bytes long.
   Data: <P "F            > 50 ED 22 46 F8 7F 00 00 00 00 00 00 00 00 00 00
  {647} normal block at 0x000001D9B291A600, 16 bytes long.
   Data: <0 "F            > 30 ED 22 46 F8 7F 00 00 00 00 00 00 00 00 00 00
  {646} normal block at 0x000001D9B291DEE0, 32 bytes long.
   Data: <                > 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00
  {645} normal block at 0x000001D9B291A010, 16 bytes long.
   Data: <  "F            > 00 ED 22 46 F8 7F 00 00 00 00 00 00 00 00 00 00
  {644} normal block at 0x000001D9B291A5B0, 16 bytes long.
   Data: <  "F            > E0 EC 22 46 F8 7F 00 00 00 00 00 00 00 00 00 00
  {643} normal block at 0x000001D9B291D1C0, 32 bytes long.
   Data: <                > 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00
  {642} normal block at 0x000001D9B291A560, 16 bytes long.
   Data: <  "F            > A8 EC 22 46 F8 7F 00 00 00 00 00 00 00 00 00 00
  {641} normal block at 0x000001D9B291AF60, 16 bytes long.
   Data: <  "F            > 88 EC 22 46 F8 7F 00 00 00 00 00 00 00 00 00 00
  {640} normal block at 0x000001D9B291D160, 32 bytes long.
   Data: <                > 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00
  {639} normal block at 0x000001D9B291A4C0, 16 bytes long.
   Data: <X "F            > 58 EC 22 46 F8 7F 00 00 00 00 00 00 00 00 00 00
  {638} normal block at 0x000001D9B291AEC0, 16 bytes long.
   Data: <8 "F            > 38 EC 22 46 F8 7F 00 00 00 00 00 00 00 00 00 00
  Object dump complete.
  Assertion failed: !memleakDetected, file C:\SDKBuild\build-X64-1.4.349.0\slangd\source\slangc\main.cpp, line 169
  abort() has been called

I'm still seeing the same thing with 2026.7, and MSVC 17.14.31


 slang-bootstrap.vcxproj -> C:\SDKBuild\build-X64-1.4.349.0\slangd\build\generators\Debug\bin\slang-bootstrap.exe
  Generating slang-core-module-generated.h, ../slang-glsl-module/slang-glsl-module-generated.h
  Compiling core module on debug build, this can take a while.
  Compiling core module took 41.00 seconds.
  Detected memory leaks!
  Dumping objects ->
  {23403823} normal block at 0x000001BF6423E2A0, 32 bytes long.
   Data: <                > 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00
  {23403822} normal block at 0x000001BF64230C40, 16 bytes long.
   Data: <(M              > 28 4D B4 A9 FF 7F 00 00 00 00 00 00 00 00 00 00
  {23403821} normal block at 0x000001BF642310A0, 16 bytes long.
   Data: < M              > 08 4D B4 A9 FF 7F 00 00 00 00 00 00 00 00 00 00
  {655} normal block at 0x000001BF64237AE0, 49 bytes long.
   Data: < m              > 88 6D 8C A9 FF 7F 00 00 01 00 00 00 00 00 00 00
  {654} normal block at 0x000001BF6423ED80, 32 bytes long.
   Data: <                > 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00
  {653} normal block at 0x000001BF6423BDE0, 16 bytes long.
   Data: < L              > C0 4C B4 A9 FF 7F 00 00 00 00 00 00 00 00 00 00
  {652} normal block at 0x000001BF6423B200, 16 bytes long.
   Data: < L              > A0 4C B4 A9 FF 7F 00 00 00 00 00 00 00 00 00 00
  {651} normal block at 0x000001BF6423E600, 32 bytes long.
   Data: <                > 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00
  {650} normal block at 0x000001BF6423BA20, 16 bytes long.
   Data: <pL              > 70 4C B4 A9 FF 7F 00 00 00 00 00 00 00 00 00 00
  {649} normal block at 0x000001BF6423BB60, 16 bytes long.
   Data: <PL              > 50 4C B4 A9 FF 7F 00 00 00 00 00 00 00 00 00 00
  {648} normal block at 0x000001BF6423EEA0, 32 bytes long.
   Data: <                > 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00
  {647} normal block at 0x000001BF6423B1B0, 16 bytes long.
   Data: < L              > 18 4C B4 A9 FF 7F 00 00 00 00 00 00 00 00 00 00
  {646} normal block at 0x000001BF6423B6B0, 16 bytes long.
   Data: < K              > F8 4B B4 A9 FF 7F 00 00 00 00 00 00 00 00 00 00
  {645} normal block at 0x000001BF6423E4E0, 32 bytes long.
   Data: <                > 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00
  {644} normal block at 0x000001BF6423B660, 16 bytes long.
   Data: < K              > C8 4B B4 A9 FF 7F 00 00 00 00 00 00 00 00 00 00
  {643} normal block at 0x000001BF6423BD90, 16 bytes long.
   Data: < K              > A8 4B B4 A9 FF 7F 00 00 00 00 00 00 00 00 00 00
  Object dump complete.
  Assertion failed: !memleakDetected, file C:\SDKBuild\build-X64-1.4.349.0\slangd\source\slangc\main.cpp, line 169
  abort() has been called
C:\Program Files\Microsoft Visual Studio\2022\Community\MSBuild\Microsoft\VC\v170\Microsoft.CppCommon.targets(254,5): error MSB8066: Custom bu
ild for 'C:\SDKBuild\build-X64-1.4.349.0\slangd\build\CMakeFiles\e3c847139e57624f67447c52895227c7\slang-core-module-generated.h.rule;C:\SDKBui
ld\build-X64-1.4.349.0\slangd\build\CMakeFiles\f2588cb960761a7f2d560851414747b1\generate_core_module_headers.rule;C:\SDKBuild\build-X64-1.4.34
9.0\slangd\source\slang-core-module\CMakeLists.txt' exited with code 3. [C:\SDKBuild\build-X64-1.4.349.0\slangd\build\source\slang-core-module
\generate_core_module_headers.vcxproj]

--- a/source/slang-record-replay/replay-context.cpp
+++ b/source/slang-record-replay/replay-context.cpp
@@ -124,27 +124,37 @@ DataMismatchException::DataMismatchException(size_t offset, size_t size)
 // ReplayContext construction and low-level helpers
 // =============================================================================
 
-// Tracks the singleton in get() once it has been constructed, so that tryGet()
-// can observe construction from other threads without racing against a
-// concurrent get() call. Guarded by s_contextMutex to match the thread-safety
-// contract documented on get().
+// Singleton pointer and its mutex. Using heap allocation instead of a
+// function-local static so that destroySingleton() can fully release the
+// instance (including its STL container internals) before _CrtDumpMemoryLeaks()
+// runs in wmain(), preventing false leak reports in MSVC Debug builds when
+// CMAKE_MSVC_RUNTIME_LIBRARY is the default MultiThreadedDebugDLL (/MDd).
 static std::mutex s_contextMutex;
 static ReplayContext* s_contextInstance = nullptr;
 
 ReplayContext& ReplayContext::get()
 {
-    static ReplayContext s_instance;
-    {
-        std::lock_guard<std::mutex> lock(s_contextMutex);
-        s_contextInstance = &s_instance;
-    }
-    return s_instance;
+    std::lock_guard<std::mutex> lock(s_contextMutex);
+    if (!s_contextInstance)
+        s_contextInstance = new ReplayContext();
+    return *s_contextInstance;
 }
 
 ReplayContext* ReplayContext::tryGet()
 {
     std::lock_guard<std::mutex> lock(s_contextMutex);
     return s_contextInstance;
+}
+
+void ReplayContext::destroySingleton()
+{
+    ReplayContext* toDelete;
+    {
+        std::lock_guard<std::mutex> lock(s_contextMutex);
+        toDelete = s_contextInstance;
+        s_contextInstance = nullptr;
+    }
+    delete toDelete;
 }
 
 ReplayContext::ReplayContext()

--- a/source/slang-record-replay/replay-context.h
+++ b/source/slang-record-replay/replay-context.h
@@ -216,14 +216,13 @@ public:
 
     /// Return the singleton if it has already been constructed, or nullptr.
     /// Thread-safe - uses a mutex to synchronize with get().
-    /// Use this in cleanup paths (e.g. slang_shutdown) to avoid constructing
+    /// Use this in cleanup paths to avoid constructing
     /// the singleton just to tear it down.
     SLANG_API static ReplayContext* tryGet();
 
-    /// Destroy the singleton and release all its heap allocations.
-    /// Called from slang_shutdown() so that _CrtDumpMemoryLeaks() does not
-    /// report the STL debug-proxy allocations owned by the singleton's
-    /// Dictionary members as leaks in MSVC Debug builds.
+    /// Return the singleton if it has already been constructed, or nullptr.
+    /// Thread-safe - uses a mutex to synchronize with get().
+    /// Use this to check whether the singleton exists without constructing it.
     SLANG_API static void destroySingleton();
 
     /// Create an idle context.

--- a/source/slang-record-replay/replay-context.h
+++ b/source/slang-record-replay/replay-context.h
@@ -220,6 +220,12 @@ public:
     /// the singleton just to tear it down.
     SLANG_API static ReplayContext* tryGet();
 
+    /// Destroy the singleton and release all its heap allocations.
+    /// Called from slang_shutdown() so that _CrtDumpMemoryLeaks() does not
+    /// report the STL debug-proxy allocations owned by the singleton's
+    /// Dictionary members as leaks in MSVC Debug builds.
+    SLANG_API static void destroySingleton();
+
     /// Create an idle context.
     /// Will switch to Record mode if SLANG_RECORD_LAYER=1 is set.
     SLANG_API ReplayContext();

--- a/source/slang-record-replay/replay-context.h
+++ b/source/slang-record-replay/replay-context.h
@@ -216,13 +216,14 @@ public:
 
     /// Return the singleton if it has already been constructed, or nullptr.
     /// Thread-safe - uses a mutex to synchronize with get().
-    /// Use this in cleanup paths to avoid constructing
+    /// Use this in cleanup paths (e.g. slang_shutdown) to avoid constructing
     /// the singleton just to tear it down.
     SLANG_API static ReplayContext* tryGet();
 
-    /// Return the singleton if it has already been constructed, or nullptr.
-    /// Thread-safe - uses a mutex to synchronize with get().
-    /// Use this to check whether the singleton exists without constructing it.
+    /// Destroy the singleton instance if it exists, freeing all resources.
+    /// Thread-safe - uses a mutex to synchronize with get() and tryGet().
+    /// Safe to call when no singleton exists (no-op). After this call,
+    /// get() will lazily re-create a fresh instance if called again.
     SLANG_API static void destroySingleton();
 
     /// Create an idle context.

--- a/source/slang/slang-api.cpp
+++ b/source/slang/slang-api.cpp
@@ -283,8 +283,7 @@ SLANG_API void slang_shutdown()
     Slang::SPIRVCoreGrammarInfo::freeEmbeddedGrammerInfo();
     Slang::RttiInfo::deallocateAll();
     Slang::freeCapabilityDefs();
-    if (auto replayContext = SlangRecord::ReplayContext::tryGet())
-        replayContext->resetHandlers();
+    SlangRecord::ReplayContext::destroySingleton();
 }
 
 SLANG_API void slang_enableRecordLayer(bool enable)

--- a/source/slang/slang.cpp
+++ b/source/slang/slang.cpp
@@ -43,6 +43,8 @@
 #include "slang-tag-version.h"
 #include "slang-type-layout.h"
 
+#include <cinttypes>
+#include <cstdio>
 #include <sys/stat.h>
 
 // Used to print exception type names in internal-compiler-error messages
@@ -58,9 +60,16 @@ const char* getBuildTagString()
         // If the tag is unknown, then we will try to get the timestamp of the shared library
         // and use that as the version string, so that we can at least return something
         // that uniquely identifies the build.
-        static String timeStampString =
-            String(SharedLibraryUtils::getSharedLibraryTimestamp((void*)spCreateSession));
-        return timeStampString.getBuffer();
+        //
+        // Use a static char buffer (BSS) instead of a heap String so MSVC
+        // _CrtDumpMemoryLeaks() at exit does not false-report that allocation.
+        static char timeStampBuffer[32] = {};
+        if (timeStampBuffer[0] == '\0')
+        {
+            uint64_t ts = SharedLibraryUtils::getSharedLibraryTimestamp((void*)spCreateSession);
+            snprintf(timeStampBuffer, sizeof(timeStampBuffer), "%" PRIu64, ts);
+        }
+        return timeStampBuffer;
     }
     return SLANG_TAG_VERSION;
 }

--- a/source/slang/slang.cpp
+++ b/source/slang/slang.cpp
@@ -45,6 +45,7 @@
 
 #include <cinttypes>
 #include <cstdio>
+#include <mutex>
 #include <sys/stat.h>
 
 // Used to print exception type names in internal-compiler-error messages

--- a/source/slang/slang.cpp
+++ b/source/slang/slang.cpp
@@ -64,11 +64,12 @@ const char* getBuildTagString()
         // Use a static char buffer (BSS) instead of a heap String so MSVC
         // _CrtDumpMemoryLeaks() at exit does not false-report that allocation.
         static char timeStampBuffer[32] = {};
-        if (timeStampBuffer[0] == '\0')
+        static std::once_flag initFlag;
+        std::call_once(initFlag, []()
         {
             uint64_t ts = SharedLibraryUtils::getSharedLibraryTimestamp((void*)spCreateSession);
             snprintf(timeStampBuffer, sizeof(timeStampBuffer), "%" PRIu64, ts);
-        }
+        });
         return timeStampBuffer;
     }
     return SLANG_TAG_VERSION;

--- a/source/slang/slang.cpp
+++ b/source/slang/slang.cpp
@@ -66,11 +66,13 @@ const char* getBuildTagString()
         // _CrtDumpMemoryLeaks() at exit does not false-report that allocation.
         static char timeStampBuffer[32] = {};
         static std::once_flag initFlag;
-        std::call_once(initFlag, []()
-        {
-            uint64_t ts = SharedLibraryUtils::getSharedLibraryTimestamp((void*)spCreateSession);
-            snprintf(timeStampBuffer, sizeof(timeStampBuffer), "%" PRIu64, ts);
-        });
+        std::call_once(
+            initFlag,
+            []()
+            {
+                uint64_t ts = SharedLibraryUtils::getSharedLibraryTimestamp((void*)spCreateSession);
+                snprintf(timeStampBuffer, sizeof(timeStampBuffer), "%" PRIu64, ts);
+            });
         return timeStampBuffer;
     }
     return SLANG_TAG_VERSION;

--- a/tools/slang-unit-test/unit-test-replay-shutdown-leak.cpp
+++ b/tools/slang-unit-test/unit-test-replay-shutdown-leak.cpp
@@ -53,3 +53,16 @@ SLANG_UNIT_TEST(replayTryGetSkipsConstruction)
     SLANG_CHECK(ptr != nullptr);
     SLANG_CHECK(ptr == &ReplayContext::get());
 }
+
+// Make sure the singleton is released after destroySingleton().
+SLANG_UNIT_TEST(replayContextDestroyedByShutdown)
+{
+    REPLAY_TEST;
+    SLANG_UNUSED(unitTestContext);
+
+    SLANG_CHECK(ReplayContext::get() != nullptr);
+
+    // Simulate what slang_shutdown() now does
+    ReplayContext::destroySingleton();
+    SLANG_CHECK(ReplayContext::tryGet() == nullptr);
+}

--- a/tools/slang-unit-test/unit-test-replay-shutdown-leak.cpp
+++ b/tools/slang-unit-test/unit-test-replay-shutdown-leak.cpp
@@ -61,6 +61,7 @@ SLANG_UNIT_TEST(replayContextDestroyedByShutdown)
     SLANG_UNUSED(unitTestContext);
 
     ReplayContext::get();
+    SLANG_CHECK(ReplayContext::tryGet() != nullptr);
 
     // Simulate what slang_shutdown() now does
     ReplayContext::destroySingleton();

--- a/tools/slang-unit-test/unit-test-replay-shutdown-leak.cpp
+++ b/tools/slang-unit-test/unit-test-replay-shutdown-leak.cpp
@@ -60,7 +60,7 @@ SLANG_UNIT_TEST(replayContextDestroyedByShutdown)
     REPLAY_TEST;
     SLANG_UNUSED(unitTestContext);
 
-    SLANG_CHECK(ReplayContext::get() != nullptr);
+    ReplayContext::get();
 
     // Simulate what slang_shutdown() now does
     ReplayContext::destroySingleton();

